### PR TITLE
Refactor to enable multi-springability

### PR DESCRIPTION
**Pull request by birneamstiel**
As discussed this PR:
*  Moves `SpringPane` reference into `UI`
*  Shifts in general more `knowledge` into the `SpringPane`, e.g. triggering of simulation runner, visualising trace, etc.
*  Makes tools and `SpringPane` use the spring constant attached to the spring links stored in the `Graph` instance

depends on #209, make-export-great-again

### CODE REVIEW

File: src/ui/dialogs/spring_pane.rb [hunk 1]
```diff
@@ -39,19 +112,54 @@ def open_dialog
...
     @dialog.show
-    register_insights_callbacks
+    register_callbacks
+  end
+
+  # compilation / simulation logic:
+
+  def try_compile
+    @simulation_runner ||= SimulationRunner.instance
+    @simulation_data ||= simulate
```
Comment (XPerianer) [inline src/ui/dialogs/spring_pane.rb]: This is double, cause simulate already writes into @simulation_data

File: src/system_simulation/simulation_runner.rb [hunk 2]
```diff
@@ -79,17 +86,20 @@ def find_equilibrium(constant = 50, mass = 20)
     raw_data.index(equilibrium_data_row)
   end
 
-  def constant_for_constrained_angle(allowed_angle_delta = Math::PI / 2.0, initial_constant = 500, mass = 20,
-                                     spring_id = 0, angle_id = 0)
+  def constant_for_constrained_angle(allowed_angle_delta = Math::PI / 2.0, spring_id = 25, initial_constant = 500)
     # steps which the algorithm uses to approximate the valid spring constant
+
+    update_spring_data_from_graph
+    angle_filter = @angles_for_springs[spring_id]
```
[the remark this answers is not in the document]
Comment (birneamstiel) [reply src/system_simulation/simulation_runner.rb]: Is there some advantage execution time wise from using `start_operation` and `commit_operation`? As far as I know this is only for creating sequential user actions to be able to use undo/redo (as stated [here](https://ruby.sketchup.com/Sketchup/Model.html#start_operation-instance_method))

File: src/system_simulation/simulation_runner.rb [hunk 3]
```diff
@@ -79,17 +86,20 @@ def find_equilibrium(constant = 50, mass = 20)
...
     step_sizes = [1500, 1000, 200, 50, 5]
```
[the remark this answers is not in the document]
Comment (birneamstiel) [reply src/system_simulation/simulation_runner.rb]: Nah not sure, I was a bit inspired by [Gradient descent](https://en.wikipedia.org/wiki/Gradient_descent) but the main idea taken from there is the selection of the step widths. I'll add a explaining comment to the method.

File: src/ui/dialogs/spring_pane.rb [hunk 4]
```diff
@@ -1,31 +1,104 @@
...
+    Graph.instance.nodes.each do | node_id, node|
+      node.update_position(current_data_sample.position_data[node_id.to_s])
+      node.hub.update_position(current_data_sample.position_data[node_id.to_s])
+      node.hub.update_user_indicator()
+    end
+
+    Graph.instance.edges.each do |_, edge|
+      link = edge.link
+      link.update_link_transformations
+    end
```
[the remark this answers is not in the document]
Comment (birneamstiel) [reply src/ui/dialogs/spring_pane.rb]: See above

File: src/ui/dialogs/spring_pane.rb (hunk 1 above)
Comment (birneamstiel) [reply src/ui/dialogs/spring_pane.rb]: Fixed in 6d9d15a

File: src/ui/dialogs/spring_pane.rb [hunk 5]
```diff
@@ -39,19 +112,54 @@ def open_dialog
     @dialog.set_html(t.result(binding))
     @dialog.set_position(500, 500)
     @dialog.show
-    register_insights_callbacks
+    register_callbacks
+  end
+
+  # compilation / simulation logic:
```
[the remark this answers is not in the document]
Comment (birneamstiel) [reply src/ui/dialogs/spring_pane.rb]: Fixed in 6d9d15a

File: src/tools/spring_tool.rb [hunk 6]
```diff
@@ -5,4 +5,9 @@ class SpringTool < ActuatorTool
   def initialize(ui)
     super(ui, 'spring')
   end
+
+  def onLButtonDown(flags, x, y, view)
+    super(flags, x, y, view)
```
[the remark this answers is not in the document]
Comment (birneamstiel) [reply src/tools/spring_tool.rb]: cool, done in 6d9d15a

File: src/tools/demonstrate_amplitude_tool.rb [hunk 7]
```diff
@@ -36,23 +38,23 @@ def onLButtonUp(_flags, x, y, view)
...
-    @trace_visualization.add_trace(['18', '20'], 4, @simulation_data)
+    spring_id = relevant_spring_id_for_node(@start_node)
+    constant = @simulation_runner.constant_for_constrained_angle(angle, spring_id)
+    @ui.spring_pane.update_constant_for_spring(spring_id, constant)
 
     view.invalidate
     reset
   end
 
+
```
[the remark this answers is not in the document]
Comment (birneamstiel) [reply src/tools/demonstrate_amplitude_tool.rb]: Fixed in 6d9d15a

File: src/system_simulation/simulation_runner.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (birneamstiel) [reply src/system_simulation/simulation_runner.rb]: Done in 6d9d15a

File: src/system_simulation/simulation_runner.rb [hunk 8]
```diff
@@ -25,11 +25,18 @@ def initialize(suppress_compilation = false, keep_temp_dir = false)
 
       run_compilation
     end
+
+    # maps spring edge id => spring constant
+    @constants_for_springs = { 21 => 7000, 25 => 7000 }
```
[the remark this answers is not in the document]
Comment (birneamstiel) [reply src/system_simulation/simulation_runner.rb]: Right you are, done in 6d9d15a

File: src/tools/demonstrate_amplitude_tool.rb [hunk 9]
```diff
@@ -80,6 +82,23 @@ def draw(view)
...
+  # Returns the spring that makes the static group the node is in movable.
+  def relevant_spring_id_for_node(node)
+    static_groups = StaticGroupAnalysis.get_static_groups_for_node(node)
+    raise 'No static groups detected' unless static_groups
+
+    static_group = static_groups[0]
+    all_spring_edges = Graph.instance.edges.values.select { |edge| edge.link_type == 'spring' }
+    spring_edges = all_spring_edges.select do |spring_edge|
+      static_group.include?(spring_edge.first_node) || static_group.include?(spring_edge.second_node)
+    end
```
[the remark this answers is not in the document]
Comment (birneamstiel) [reply src/tools/demonstrate_amplitude_tool.rb]: Yep, better that way... done in 4a39287

**Pull request merged**